libs/langchain/langchain/vectorstores/elastic_vector_search.py

```diff
@@ -608,12 +589,167 @@ def knn_hybrid_search(
         }
 
         # Perform the hybrid search on the Elasticsearch index and return the results.
-        res = self.client.search(
+        response = self.client.search(
             index=self.index_name,
             query=match_query_body,
             knn=knn_query_body,
             fields=fields,
             size=size,
             source=source,
         )
-        return dict(res)
+
+        hits = [hit for hit in response["hits"]["hits"]]
+        docs_and_scores = [
+            (
+                Document(
+                    page_content=hit["_source"][page_content]
+                    if source
+                    else hit["fields"][page_content][0],
+                    metadata=hit["fields"] if fields else {},
+                ),
+                hit["_score"],
+            )
+            for hit in hits
+        ]
+
+        return docs_and_scores
+
+    def create_knn_index(self, mapping: Dict) -> None:
+        """
+        Create a new k-NN index in Elasticsearch.
+
+        Args:
+            mapping (Dict): The mapping to use for the new index.
+
+        Returns:
+            None
+        """
+
+        self.client.indices.create(index=self.index_name, mappings=mapping)
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[Dict[Any, Any]]] = None,
+        model_id: Optional[str] = None,
+        refresh_indices: bool = False,
+        **kwargs: Any,
+    ) -> List[str]:
+        """
+        Add a list of texts to the Elasticsearch index.
+
+        Args:
+            texts (Iterable[str]): The texts to add to the index.
+            metadatas (List[Dict[Any, Any]], optional): A list of metadata dictionaries to associate with the texts.
+            model_id (str, optional): The ID of the model to use for transforming the texts into vectors.
+            refresh_indices (bool, optional): Whether to refresh the Elasticsearch indices after adding the texts.
+            **kwargs: Arbitrary keyword arguments.
+
+        Returns:
+            A list of IDs for the added texts.
+        """
+
+        try:
+            from elasticsearch.helpers import bulk
+        except ImportError:
+            raise ImportError(
+                "Could not import elasticsearch python package. "
+                "Please install it with `pip install elasticsearch`."
+            )
+
+        # Check if the index exists.
+        if not self.client.indices.exists(index=self.index_name):
+            dims = kwargs.get("dims")
```

Comment:
is dims something that could be stored on the class? does it change from one call of add_texts to the next for a single vector store object?

Reply:
In `add_texts` `dims` is only checked and required after confirming an index does not exist (so an index can be created with the correct mapping). 
Normally you would only pass `dims` the one time you want to create an index either through `add_texts` or `from_texts`